feat: Add support for custom plan executors

OlegWock opened the pull request:
Hello Jon! We previously had a [discussion](https://github.com/vega/vegafusion/discussions/567) about integrating VegaFusion with distributed query executors. I wanted to express my thanks once again for your help. I was able to marry VegaFusion with Spark and we're currently using it in production. This involved adding support for custom query/plan executors in VF and I thought this might be something worth contributing back upstream.

This PR turned out to be quite big, so I'll put here summary of the changes and will also add couple of comments to specific pieces of code.

## Main changes

1. Updated VF to not materialize DataFrame until necessary. Previously dataframes were materialized after each transformation step, now VF will try to preserve them as logical plan without execution when possible. For most of tested cases they can be materialized just once, right before emedding into transformed spec.
    * This required adding new variant `Plan` to `TaskValue` enum to allow passing plans between tasks.
    * In some contexts we expect only task value with materialized data (i.e. scalar or table), so for these cases new `MaterializedTaskValue` was introduced with only `Scalar` and `Table` variants (+ methods to convert between two enums).
    * Similarly, there is now separate `ExportUpdate` (might include plan) and `ExportUpdateArrow` (materialized) enums

2. Added support for custom plan executors. `PlanExecutor` is an object which receives DataFusion's logical plan and returns the data. This is main point for connecting VF to other query engines (we used it for Spark, but it should be possible to use it also for Snowpark or BigQuery, for example). I also included `DataFusionPlanExecutor` which uses DataFusion to execute the plan, which is default and preserves old behavior.
    * Custom executors aren't exposed to Python API (available only in Rust). I tried both protobuf and using `datafusion-python` crate directly, but I couldn't make serialization of LogicalPlan work. Error was something to do with TableSource not being serializable, I don't remember exactly, it was past my understanding of DF unfortunately. I can provide more details if you wish.

3. Used `Plan` variant of `VegaFusionDataset` to allows to not bring any actual data as input and allow VF to build transformations on top of plan instead. Currently, this is useful only in combination with custom plan executor (as otherwise DataFusion won't have the means to execute initial plan/get input data). From user perspective, this is implemented as an option to pass Arrow's `Schema` as inline dataset. It will be translated into simple `TableScan` logical plan node and wrapped in `VegaFusionDataset::Plan`

4. Expression compilation is made async. Since some expression operated on materialized data, with introduction of custom executor there is a possibility that data is not there when expression needs it and thus it needs to be materialized first. Since materialization is async function, whole compilation chain had to become async as well.

5. I updated DataFusion crate (to v49.0, as updating further required updating bunch of other packages and it didn't end well last time I tried 😅) and fixed some other (unrelated really) issues here and there. I'll add comments to specific pieces of code.

---

What I'm not very confident is WASM/gRPC side of the project. As I said we're already using this in production for some time, but we're only using Python bindings. I tried updating it accordingly, but I don't fully understand how I would test it. I would appreciate if you could take a look advice how to test it.ys

OlegWock commented:
Yes, that was my intention. After this PR is merged I'll need to reconcile our local fork with upstream (I didn't do the best job at keeping clean git history 😅), but all changes we have there are kind of "superset" of this PR so it should come together nicely

OlegWock commented:
@jonmmease I merged upstream into this branch and resolved conflicts. Could you approve workflow runs?
And please let me know if I can help get this merged in any other way

jonmmease commented:
Thanks @OlegWock, and sorry for the delays here. I'll review this week

jonmmease commented:
An early thought here, do you have thoughts on how fallback should work if a PlanExecutor doesn't support some feature of the logical plan?  The most common example that comes up is `median` aggregation, which isn't supported by all SQL databases.

Do we want this to just be an error if the executor doesn't support it? This would have the most predictable performance, but wouldn't provide full coverage of valid Vega specs.  Or do we want to fall back to DataFusion? Or have the PlanExecutor broadcast the features it supports so the planner can avoid planning unsupported operations on the server?

(edit): I guess the fallback logic could be handled by something like [datafusion-contrib/datafusion-federation](https://github.com/datafusion-contrib/datafusion-federation), which could implement the PlanExecutor trait

jonmmease commented on examples/rust-examples/examples/custom_executor.rs:
nice, thanks for adding the usage example

jonmmease commented on vegafusion-core/src/runtime/plan_executor.rs:
```diff
@@ -0,0 +1,23 @@
+
+#[async_trait]
+pub trait PlanExecutor: Send + Sync {
+    async fn execute_plan(&self, plan: LogicalPlan) -> Result<VegaFusionTable>;
```
I'd lean toward adding a name as well

```
fn name(&self) -> &str 
```

just so VegaFusion has some identifier it can use in the future for logging etc.

OlegWock replied:
Done ✔️

jonmmease commented on vegafusion-core/src/task_graph/memory.rs:
```diff
@@ -94,3 +95,12 @@ pub fn inner_size_of_table(value: &VegaFusionTable) -> usize {
     schema_size + size_of_batches
 }
+
+pub fn inner_size_of_logical_plan(plan: &LogicalPlan) -> usize {
+    size_of_val(plan)
+        + plan
+            .inputs()
+            .iter()
+            .map(|p| inner_size_of_logical_plan(p))
+            .sum::<usize>()
```
yeah, this looks fine for now

jonmmease commented on vegafusion-core/src/task_graph/task_value.rs:
I like this formulation of having `TaskValue` for potentially lazy values and `MaterializedTaskValue` for the materialized version

jonmmease commented on vegafusion-core/src/task_graph/task_value.rs:
```diff
@@ -75,11 +122,34 @@ impl TryFrom<&TaskValue> for ProtoTaskValue {
+            }),
+            TaskValue::Plan(_) => Err(VegaFusionError::internal(
+                "TaskValue::Plan cannot be serialized to protobuf. Plans are intermediate values that should be materialized to tables using .to_materialized(plan_executor) before serialization.",
+            )),
```
I don't think we need this today, but DataFusion's logical plans are serializable through [datafusion-proto](https://docs.rs/crate/datafusion-proto/latest)

OlegWock replied:
I tried implementing serialization, but couldn't get it working. My memory might be a bit blurry at this point, but as I recall I tried both making custom `TableSource` (or maybe `TableProvider`) as well as using `EmptyTable`. But both were failing with `TableSource was not DefaultTableSource` error (possibly related issue https://github.com/apache/datafusion/issues/16749). I believe this was also the reason why custom executor API is not exposed to Python, as there were no way to transfer plan to Python

jonmmease commented on vegafusion-core/src/task_graph/task.rs:
Do you think we could change value here to be `MaterializedTaskValue` since Plan isn't valid here?

OlegWock replied:
Done ✔️

jonmmease commented on vegafusion-core/src/runtime/runtime.rs:
```diff
@@ -347,28 +375,95 @@ pub trait VegaFusionRuntimeTrait: Send + Sync {
+        ChartSpec,
+        Vec<ExportUpdate>,
+        Vec<PreTransformLogicalPlanWarning>,
+    )> {
```
looks like we're not using this internally at this point, is this the entry point you're are using to wire the `LogicalPlans` in these `ExportUpdate`s to Spark?

Having this method would push me towards making the `Plan` variant of `TaskValue` in ExportUpdate serializable, so this could be served over gRPC like the rest of the `pre_transform_*` methods at this level

OlegWock replied:
> is this the entry point you're are using to wire the LogicalPlans in these ExportUpdates to Spark

No, for Spark we use `pre_transform_spec` (but construct VF runtime with custom executor). I just though this API (`pre_transform_logical_plan`) might be useful to some, so I added a new public method, since most of the code was already there, just not exposed

> Having this method would push me towards making the Plan variant of TaskValue in ExportUpdate serializable, so this could be served over gRPC like the rest of the pre_transform_* methods at this level

I'd be happy to help with this, but as far as I understand, currently there is no way to serialize plan we construct (see my other comment)

jonmmease commented on vegafusion-core/src/planning/watch.rs:
```diff
@@ -120,13 +120,21 @@ impl TryFrom<VariableNamespace> for ExportUpdateNamespace {
+    pub name: String,
+    pub scope: Vec<u32>,
+    pub value: MaterializedTaskValue,
+}
```
Could we call this `MaterializedExportUpdate`?

OlegWock replied:
Done ✔️

jonmmease commented on vegafusion-core/src/runtime/runtime.rs:
```diff
@@ -40,13 +41,37 @@ pub struct PreTransformExtractTable {
+    fn plan_executor(&self) -> Arc<dyn PlanExecutor> {
+        Arc::new(NoOpPlanExecutor)
+    }
+
```
I think I'd rather not expose the PlanExector through the trait, which is causing the need for the `NoOpPlanExecutor` to handle the grpc and wasm configurations. Could we instead add a new trait method

```
async fn materialize(&self, value: TaskValue) -> Result<MaterializedTaskValue>;
```

and have the other implemented trait methods here call that?  We can still have a default implementation for now that Errors on LogicalPlans, but this would set us up well to eventually serialize the LogicalPlans through protobuf as well.

OlegWock replied:
Done ✔️

I changed a signature slightly to accept vector instead of single value. This makes it easier to then re-define parallel materialization in vegafusion-runtime implementation (and also matches other method `materialize_export_updates` on trait)

jonmmease commented on vegafusion-core/src/proto/pretransform.proto:
```diff
@@ -114,4 +114,24 @@ message PreTransformExtractRequest {
+  string name = 2;
+  repeated uint32 scope = 3;
+  tasks.MaterializedTaskValue value = 4;
+}
```
Could we name this `MaterializedExportUpdate`?

OlegWock replied:
Done ✔️

jonmmease commented on vegafusion-runtime/src/data/tasks.rs:
```diff
@@ -179,43 +195,36 @@ impl TaskCall for DataUrlTask {
+            let pipeline = self.pipeline.as_ref().unwrap();
+            pipeline.eval_sql(df, &config).await?
+        } else {
+            // No transforms, just remove any ordering column
```
is this comment stale?

OlegWock replied:
Yes, removed

jonmmease commented on vegafusion-runtime/src/data/tasks.rs:
```diff
@@ -131,19 +143,23 @@ impl TaskCall for DataUrlTask {
+        let inline_dataset_info: Option<&VegaFusionDataset> = inline_name
+            .as_ref()
+            .and_then(|name| inline_datasets.get(name));
+
```
In your spark usecase, are you passing in inline logical plans representing tables or queries? Or passing in csv/parquet urls?

OlegWock replied:
Not sure if I get your question right, but for Spark we pass a Arrow Schema from Python, which is transformed into logical plan on Rust side

jonmmease commented on vegafusion-runtime/src/data/tasks.rs:
```diff
@@ -179,43 +195,36 @@ impl TaskCall for DataUrlTask {
+            .unwrap_or(false)
+        {
+            let pipeline = self.pipeline.as_ref().unwrap();
+            pipeline.eval_sql(df, &config).await?
```
There are cases where it would be a performance regression for the local DataFusion PlanExecutor to not evaluate Plans to Tables at the end of the pipeline of transforms (when these are used repeatedly in interactive scenarios). So I think we'll need some ability to control this.

To be conservative to start, could we set things up so that when using the `DataFusionPlanExecutor` we evaluate these DataFrames to VegaFusionTable at the end of the pipeline? Then for any other PlanExecutor, leave them lazy.  Later we can put more thought into whether the PlanExecutor trait should broadcast a preference somehow, but we don't need to figure that out now.

OlegWock replied:
Sure, I can do that. But in this case it will make previously mentioned `pre_transform_logical_plan` useless when used with default DF executor, should I remove the `pre_transform_logical_plan ` then?

jonmmease reviewed (commented):
I haven't gotten all the way through, but wanted to share initial feedback. This is looking great!  I have some comments below, see what you think.

I still need to look at the python, server, wasm, and test changes, and will aim to get to those tomorrow

jonmmease commented on vegafusion-python/src/utils.rs:
```diff
@@ -0,0 +1,96 @@
+                        VegaFusionDataset::from_table(table, Some(hash))?
+                    } else if let Ok(pyschema) = inline_dataset.extract::<PySchema>() {
+                        // Handle PyArrow Schema as VegaFusionDataset::Plan
+                        let schema = pyschema.into_inner();
```
Are you currently using the Python API this way, passing in PySchema to represent external spark tables?  Without the ability to define a custom plan executor in Python, I don't quite follow how you could make use of this.

The general idea of representing external tables by a schema that get's turned into an EmptyTable and then filled in by the PlanExecutor later is a nice approach, I'm just not sure it's usable in this PR as-is

OlegWock replied:
Yes, we use this API. You're right that it's not very usable at this point because there is no Python bindings for custom executor API. I wasn't able to get plan to be serialized to send it to Python, I tried both serializing into protobuf (I mention error in one of other comments) and using `datafusion_python` crate to transform LogicalPlan from DF into Python wrapper object directly, but it didn't work either (not saying it's impossible, it's skill issue more likely than not). 

But I think it's nice future-proof API and it made sense to me to include it in this PR.

---

If you're interested in how we use it, it's a bit hacky but we have custom [`SparkSqlPlanExecutor`](https://github.com/deepnote/vegafusion/blob/main/vegafusion-python/src/vendor.rs#L17) which accepts Python function as an argument, and when it [needs to execute DataFusion LogicalPlan](https://github.com/deepnote/vegafusion/blob/main/vegafusion-python/src/vendor.rs#L71), it converts it to SQL string, which is then send to Python for execution (as that's where available Spark client lives) and Python returns Arrow table to Rust. Here is [example usage](https://github.com/deepnote/deepnote-toolkit/blob/main/deepnote_toolkit/chart/deepnote_chart.py#L33-L50) from Python side. This way we were able to avoid issues with serializing DataFusion plans. But it's rather ad-hoc solution for our specific problem, so I didn't think it's worth including any parts of it in this PR.

jonmmease replied:
Ok, does this mean that you'll continue to need to fork `vegafusion-python` until we figure out plan serialization so that you can define a Plan executor in Python?

OlegWock replied:
I think we'll stick with our current approach for foreseeable future. For our specific case, even if/when we solve serialization issue, we already have implemented LogicalPlan -> Spark SQL converter in Rust. Because conversion is not straightforward, there are a lot of rewrites and patching for both LogicalPlan before it's converted to SQL AST, and to SQL AST before it's stringified to make it compatible with Spark. It does the job well now, so it won't be our priority to port it to Python (to be able to use VF without forking). So I think we'll just keep using our fork, periodically rebasing on top of upstream

jonmmease replied:
That makes sense. Will you be able to at least point your `vegafusion-python` fork to the upstream `vegafusion-*` crates once we merge and release this?

OlegWock replied:
I'm not sure I understand what you mean or how it works. Do you mean a case when our fork will contain only modified vegafusion-python, and then import vegafusion-runtime/-core/-etc installed from crates.io instead of having source directly in the repository?

jonmmease replied:
Yes, that's what I was curious about

OlegWock replied:
It will require moving Spark-related code to vegafusion-python (as I initially put it into runtime), but besides that it should be possible, yes

OlegWock commented:
I addressed most of the comments, one remaining is whether I should just remove `pre_transform_logical_plan` due to its potential incompatibility with DataFusion executor if it's made to be eager instead of lazy (to better mirror existing behavior and avoid performance drop for interactive workflows)

jonmmease commented:
Thanks @OlegWock, let's remove `pre_transform_logical_plan` from this PR since you're not using it and switch to evaluating DataFrames to `VegaFusionTable` at the end of transform pipelines when using `DataFusionPlanExecutor`.  This will preserve prior DataFusion performance characteristics. I'll look into LogicalPlan serialization, as I had it working at one point in the VegaFusion 1 series, but with these changes I don't think we need it for this PR.

OlegWock commented:
Sure, I removed it and added materialization in between pipelines for plans when using DF executor

jonmmease commented:
Thanks, I'll give this one more look shortly

For future reference, I looked back at v1 and realized that we were only serializing `Expr`s to protobuf, not entire logical plans.

https://github.com/vega/vegafusion/blob/30f91b89a15a7a1b86ea5064ff15803166c1c92b/vegafusion-python-embed/src/dataframe.rs#L524-L541

my hope would be to get back to something along these lines, where the Rust PyO3 python code serializes the logical plan to protobuf, then parses into the Python object form (using the Python generated by protoc), then passes that to a Python PlanExecutor.  That way a custom executor could be written entirely in Python, and implemented against the Python protoc wrappers around the LogicalPlan.

OlegWock commented:
Thanks.

Regarding serialization, I'll take a look this or next week, maybe I'll be able to come up with something with a fresh head.

OlegWock commented:
I took another look on serialization as promised. With some mixed results.

First of all, I realized I gave you wrong error (`TableSource was not DefaultTableSource`) in previous comment. Actual error was `LogicalExtensionCodec is not provided` when trying to serialize a plan. It is caused by the fact that default codec can't handle `EmptyTable` which we use to construct plans.

One option would be to provide custom codec, but that would work only for Rust, as DF Python bindings don't provide a way to use custom codec when unserializing.

I tried rewriting plan before serialization to swap `EmptyTable` with `ListingTable` (which can be serialized) pointing to dummy file and then swap back after deserialization (to keep plan tidy), and that worked quite well for Rust, but Python bindings of DF don't expose needed methods to rewrite plan. Claude suggested converting plan to sql and then back which surprisingly seem to work, though is very hacky. Here are relevant [rust](https://github.com/deepnote/vegafusion/blob/oleh/plan-serialization-exp/vegafusion-runtime/tests/test_plan_serialization.rs) and [python](https://github.com/deepnote/vegafusion/blob/oleh/plan-serialization-exp/vegafusion-runtime/tests/test_plan_deserialization.py) files if you want to take a look.

All in all, I think best approach would be to contribute `EmptyTable`  serialization/deserialization support to datafusion itself. Seem like reasonable feature, so I would assume they won't be against it. In this case VF would be able to just serialize plan as any normal logical plan and python side would be able to deserialize it. I'll check what it would take to add that support and maybe create an issue/PR in datafusion next week.

jonmmease commented:
thanks for investigating, yeah, contributing EmptyTable serialization to DataFusion sounds like a good path. Would you mind opening that as an issue over there? With the refactoring you done, this isn't blocking for this PR, but it would be great to have a path documented for how we can get to supporting custom executors in Python. I'm planning to go over this once more tomorrow and merge it.

OlegWock commented:
Sure, I'll take it from here, don't worry 🫡

Thanks again for the review
